### PR TITLE
Issue #301: Add property-based enemy randomization invariant tests

### LINKED ISSUE #301: Add property-based tests for enemy randomization invariants
## Summary
Add property-based tests for enemy copy-ability randomization invariants across many seeds/event keys.

## Current State (2026-04-13)
The codebase now has a strong example-based suite in worlds/kirbyam/test/test_enemy_copy_ability_randomization.py covering:
- Whitelist/validation constraints (including forbidden abilities such as Wait)
- Deterministic contracts for shuffled and completely-random modes
- Weighting behavior for no-ability outcomes
- Deterministic shuffled assignment coverage helper behavior

However, there is still no property-based framework usage (for example, no Hypothesis @given tests), so broad randomized invariant coverage is still missing.

## Scope
- Add property-based tests for policy generation and lookup helpers in worlds/kirbyam/ability_randomization.py:
  - build_enemy_copy_ability_policy
  - ability_for_enemy_type
  - ability_for_enemy_grant_event
  - build_shuffled_enemy_type_assignments
- Validate whitelist-only outputs for generated policies and lookups.
- Validate forbidden abilities (notably Wait) are never emitted from valid policies.
- Validate deterministic behavior contracts over many generated seeds/keys:
  - Shuffled mode: stable per enemy type key for a fixed policy.
  - Completely random mode: stable per (enemy type key, event index) and key-sensitive across differing event inputs.
- Validate no-ability weighting invariants at boundary values (0 and 100) and sanity monotonic behavior in sampled runs.

## Why
Current tests are comprehensive but mostly example-based. Property-based testing increases confidence against edge-case regressions across large input spaces.

## Acceptance Criteria
- Property-based tests added under worlds/kirbyam/test/ (new file or extension of existing enemy randomization tests).
- CI executes these tests reliably in the standard unit test workflow.
- Failures include actionable diagnostics with reproducible seed/input context.
- Existing example-based tests remain and continue to pass.

## Revalidation (2026-04-13)
Status: Still relevant. Example-based enemy randomization tests are in place, but property-based invariant coverage across generated seeds/keys is still not present.

### PR DESCRIPTION
**Pull request by hasherwi**
## Summary
- add Hypothesis-backed property tests for enemy copy-ability randomization invariants
- cover shuffled and completely-random deterministic contracts across generated seeds/keys
- validate whitelist-preserving outputs, forbidden-ability exclusion, and no-ability boundary behavior
- validate shuffled assignment whitelist coverage when key count permits

## Files Changed
- ci-requirements.txt
- worlds/kirbyam/test/test_enemy_copy_ability_randomization_properties.py

## Validation
- python -m pytest worlds/kirbyam/test/test_enemy_copy_ability_randomization.py worlds/kirbyam/test/test_enemy_copy_ability_randomization_properties.py
- result: 24 passed

## Notes
- no slot_data changes required
- no changelog update needed (test-only hygiene coverage for unreleased scope)

Closes #301

### CODE REVIEW

File: worlds/kirbyam/test/test_enemy_copy_ability_randomization_properties.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
+"""Property-based tests for enemy copy-ability randomization invariants (Issue #301)."""
+# mypy: disable-error-code=untyped-decorator
+
+from __future__ import annotations
+
+import random
+
+from hypothesis import given, settings, strategies as st  # type: ignore[import-not-found]
```
[the remark this answers is not in the document]
Comment (hasherwi) [reply worlds/kirbyam/test/test_enemy_copy_ability_randomization_properties.py]: Implemented in 5a41f6d6: this module now uses pytest.importorskip("hypothesis") with a clear skip reason, so environments without Hypothesis skip cleanly instead of failing collection.
